sdk/trace/span_benchmark_test.go

```diff
@@ -0,0 +1,68 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+)
+
+func attributes(n int) []attribute.KeyValue {
+	a := make([]attribute.KeyValue, n)
+	for i := 0; i < n; i++ {
+		a[i] = attribute.Int(fmt.Sprint(i), i)
+	}
+	return a
+}
+func benchmarkSetAttributes(b *testing.B, i int) {
+	attrs := attributes(i)
+
+	s := &span{
+		startTime:  time.Now(),
+		attributes: newAttributesMap(i),
+	}
```

Comment:
I think that calling this via the struct directly can be optimized in a way that it can't when called via the `trace.Span` interface that users directly interact with. I can see different results changing that.


```suggestion
	var s *span

	s = &span{
		startTime:  time.Now(),
		attributes: newAttributesMap(128),
	}}
```

Reply:
If you take another look I added a number of additional benchmarks.  The interface only has a performance impact when the AttributeMap is smaller than the number of attributes added.

But even still I don't see any performance impact of a `WithAttributes` using a slice of 1, vs a single attribute.

```
go test -bench Span_Set .
goos: linux
goarch: amd64
pkg: go.opentelemetry.io/otel/sdk/trace
cpu: Intel(R) Xeon(R) Platinum 8168 CPU @ 2.70GHz
BenchmarkSpan_SetAttributes_1-4                          8371192               148.8 ns/op            64 B/op          1 allocs/op
BenchmarkSpan_SetAttributes_10-4                          865162              1458 ns/op             640 B/op         10 allocs/op
BenchmarkSpan_SetAttributes_100-4                          74750             14998 ns/op            6400 B/op        100 allocs/op
BenchmarkSpan_SetAttributes_1000-4                          7770            161611 ns/op           64022 B/op       1000 allocs/op
BenchmarkSpan_SetAttributes_Interface_1-4                8111154               145.1 ns/op            64 B/op          1 allocs/op
BenchmarkSpan_SetAttributes_Interface_10-4               1000000              1406 ns/op             640 B/op         10 allocs/op
BenchmarkSpan_SetAttributes_Interface_100-4                31867             38103 ns/op           11200 B/op        200 allocs/op
BenchmarkSpan_SetAttributes_Interface_1000-4                2742            422945 ns/op          122368 B/op       2027 allocs/op
BenchmarkSpan_SetAttribute_Interface-4                   8377492               137.5 ns/op            64 B/op          1 allocs/op
BenchmarkSpan_SetAttribute_FullMap-4                     8656261               143.3 ns/op            64 B/op          1 allocs/op
BenchmarkSpan_SetAttribute_Interface_FullMap-4           8247140               143.0 ns/op            64 B/op          1 allocs/op
PASS
ok      go.opentelemetry.io/otel/sdk/trace      15.193s
```

Reply:
Did you push this commit @MadVikingGod, can't see them.

Reply:
No I did not.  It's up now.